crates/scmi/src/scmi.rs

```diff
@@ -0,0 +1,1380 @@
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.ret_bytes.len()
+    }
+
+    pub(crate) fn communication_error(&self) -> Self {
+        Self::from(self.header, Response::from(ReturnStatus::CommsError))
+    }
+}
+
+/// Representation of a parsed SCMI request.
+///
+/// Use [ScmiRequest::get_unsigned] and [ScmiRequest::get_usize] functions to
+/// retrieve its parameters as `u32` and `usize` values respectively.
+pub struct ScmiRequest {
+    header: MessageHeader,     // 32-bit unsigned integer, split below:
+    message_id: MessageId,     // bits 7:0
+    message_type: MessageType, // bits 9:8
+    protocol_id: ProtocolId,   // bits 17:10
+    // token: u16,             // bits 27:18
+    // bits 31:28 are reserved, must be 0
+    parameters: Option<MessageValues>, // set later based on the number of parameters
+}
+
+impl ScmiRequest {
+    pub(crate) fn new(header: MessageHeader) -> Self {
+        let protocol_id: u8 = ((header >> 10) & 0xFF).try_into().unwrap();
+        let message_id: u8 = (header & 0xFF).try_into().unwrap();
+        // Token is an arbitrary info, the Linux SCMI driver uses it as a sequence number.
+        // No actual meaning for vhost except copying the unchanged header in the response
+        // as required by SCMI specification. We extract it here only for debugging purposes.
+        let token: u16 = ((header >> 18) & 0x3FF).try_into().unwrap();
+        let message_type = match (header >> 8) & 0x3 {
+            0 => MessageType::Command,
+            _ => MessageType::Unsupported,
+        };
+        debug!(
+            "SCMI request: protocol id={}, message id={}, message_type={:?}, token={}",
+            protocol_id, message_id, message_type, token
+        );
+        Self {
+            header,
+            message_id,
+            message_type,
+            protocol_id,
+            parameters: None,
+        }
+    }
+
+    fn get_unsigned(&self, parameter: usize) -> u32 {
+        self.parameters.as_ref().expect("Missing parameters")[parameter].get_unsigned()
+    }
+
+    fn get_usize(&self, parameter: usize) -> usize {
+        self.get_unsigned(parameter) as usize
+    }
+}
+
+enum ParameterType {
+    _SignedInt32,
+    UnsignedInt32,
+}
+type ParameterSpecification = Vec<ParameterType>;
+
+type HandlerFunction = fn(&ScmiHandler, &ScmiRequest) -> Response;
+
+/// Specification of an SCMI message handler.
+///
+/// No need to create this directly, use [HandlerMap::bind] to add message
+/// handlers.
+struct HandlerInfo {
+    name: String,
+    parameters: ParameterSpecification,
+    function: HandlerFunction,
+}
+
+/// Mapping of SCMI protocols and messages to handlers.
+///
+/// See [HandlerMap::bind] and [make_handlers] how to add new handlers.
+// HandlerMap layout is suboptimal but let's prefer simplicity for now.
+struct HandlerMap(HashMap<(ProtocolId, MessageId), HandlerInfo>);
+
+impl HandlerMap {
+    fn new() -> Self {
+        Self(HashMap::new())
+    }
```

Comment:
Why not derive(Default)?

Reply:
This is not intended to have a default value really. It's a constructor and it may become not completely trivial in future.